File: go/internal/sol/p0087/p0087.go
package p0087

import (
	"slices"
	"strconv"

	"pe-solver/internal/mylib"
)

func compute(limit int) string {
	type intSet map[int]struct{}
	nSet := intSet{}
	sqLst := mylib.Primes(mylib.Isqrt(limit - mylib.Pow(2, 3) - mylib.Pow(2, 4)))
	cbLst := slices.DeleteFunc(slices.Clone(sqLst), func(x int) bool { return mylib.Pow(x, 3) > limit })
	quLst := slices.DeleteFunc(slices.Clone(cbLst), func(x int) bool { return mylib.Pow(x, 4) > limit })

	for _, z := range quLst {
		for _, y := range cbLst {
			for _, x := range sqLst {
				if tmp := mylib.Pow(x, 2) + mylib.Pow(y, 3) + mylib.Pow(z, 4); tmp < limit {
					nSet[tmp] = struct{}{}
				}
			}
		}
	}

	return strconv.FormatInt(int64(len(nSet)), 10)
}

func Solve() string {
	return compute(50_000_000)
}

File: haskell/src/Sol/P0087.hs
module Sol.P0087 (compute, solve) where

{-
This implementation is quite slow. The following is a result on my Raspberry Pi 4.
I have no ideas to improve performance so far.

% cabal v2-run pe-solver -- 87
[Problem 87]
Answer: 1097343
Elapsed time: 5.112869 sec.
-}

import qualified Data.IntSet as S (fromList, size)

import Mylib.Prime (primeNumbers)

compute :: Int -> String
compute limit =
    show
        . S.size
        . S.fromList
        . filter (< limit)
        $ [ x2 + y3 + z4
          | x2 <- sqLst
          , y3 <- cbLst
          , z4 <- quLst
          ]
  where
    pow :: Int -> Int -> Int
    pow b e = b ^ e

    sqLst = takeWhile (< limit) . map (`pow` 2) $ primeNumbers
    cbLst = takeWhile (< limit) . map (`pow` 3) $ primeNumbers
    quLst = takeWhile (< limit) . map (`pow` 4) $ primeNumbers

solve :: String
solve = compute 50_000_000
